### PR TITLE
Have FileSystemError adopt ErrorNoTelemetry

### LINKED ISSUE #155492: Unable to read file errors are being localized and showing up in telemetry
https://github.com/microsoft/vscode/blob/8f58d9df260144d72545f66a964e88232d751f5f/src/vs/platform/files/common/fileService.ts#L569-L574

This shows up in telemetry: `无法读取文件`

Should `FileOperationError` extend `NoErrorTelemetry`?

https://vscode-errors.azurewebsites.net/card?ch=f3528d481b366b24321c382d5e5849f4c7d65248&bH=aa9e18ff-8c8f-1c63-d265-d1d3cb2ba855
https://vscode-errors.azurewebsites.net/card?ch=f3528d481b366b24321c382d5e5849f4c7d65248&bH=749131eb-6952-232e-4533-df818832582b

### PR DESCRIPTION
**Pull request by lramos15**
Fixes #155492

Allow setting error name with `ErrorNoTelemetry`. Tested on throwing a filesystem error but it presents a modal without any notion of ErrorNoTelemetry. If anyone has a notifcation example I can also test making sure it doesn't show up there.

### CODE REVIEW

**Comment by sbatten** [comment]
@lramos15 does it have ENOPRO in the dialog?

**Comment by lramos15** [comment]
> @lramos15 does it have ENOPRO in the dialog?

I didn't test ENOPRO specifically I tested the error listed in #155492 but the modal just had the message and not the name

File: src/vs/base/common/errors.ts [hunk 1]
```diff
@@ -258,7 +265,7 @@ export class ErrorNoTelemetry extends Error {
 	}
 
 	public static isErrorNoTelemetry(err: Error): err is ErrorNoTelemetry {
```
Comment (bpasero) [inline src/vs/base/common/errors.ts:267]: Can we implement this method differently without having to change the name of the error, e.g. by adding another property that is private to the error?

File: src/vs/base/common/errors.ts [hunk 2]
```diff
@@ -239,11 +239,18 @@ export class ExpectedError extends Error {
...
 	constructor(msg?: string) {
 		super(msg);
-		this.name = 'ErrorNoTelemetry';
+	}
+
+	public override set name(newName: string) {
+		this._name = `ErrorNoTelemetry${newName}`;
+	}
+
+	public override get name(): string {
```
Comment (bpasero) [inline src/vs/base/common/errors.ts:252]: My worry with this change is that we might have existing code that probes on `error.name` to figure out the file service error:

image

(above is a "Find Referenfes" on `error.name`).

For example here:

https://github.com/microsoft/vscode/blob/13ba7bb446a638d37ebccb1a7d74e31c32bb9790/src/vs/workbench/api/common/extHostFileSystemConsumer.ts#L149-L164

and

https://github.com/microsoft/vscode/blob/13ba7bb446a638d37ebccb1a7d74e31c32bb9790/src/vs/workbench/api/browser/mainThreadFileSystem.ts#L140-L162

In other words, this may change how extensions can detect which error was thrown?

**Review by bpasero — COMMENTED** [review]
This may be user facing in certain cases, so what is the reason for making the error name rather cryptic by concatenating `ErrorNoTelemetry` and the original `name`? I would rather use something that is more generic, since I believe the main intent is to implement `isErrorNoTelemetry` properly: 

```ts
public override set name(newName: string) {
	this._name = `${newName} (CodeExpectedError)`;
}

public static isErrorNoTelemetry(err: Error): err is ErrorNoTelemetry {
	return err.name.contains('(CodeExpectedError)');
}
```

I think `CodeExpectedError` can be something else too, if we want to go with node.js notion, maybe `EEXPECTED`).

File: src/vs/base/common/errors.ts (hunk 2 above)
Comment (lramos15) [reply src/vs/base/common/errors.ts:252]: We tried that but found when errors got serialized across processes that it would be only the base error properties that were left and therefore we had to use one of the three error properties and figured name was the least important one. @sbatten and I have tried many solutions but our architecture is quite complex so ends up breaking the ErrorNoTelemetry one way or another. It seems the name probing is unique to FileSystemErrors

File: src/vs/base/common/errors.ts (hunk 1 above)
Comment (lramos15) [reply src/vs/base/common/errors.ts:267]: We tried that but found when errors got serialized across processes that it would be only the base error properties that were left and therefore we had to use one of the three error properties and figured name was the least important one. @sbatten and I have tried many solutions but our architecture is quite complex so ends up breaking the ErrorNoTelemetry one way or another. It seems the name probing is unique to FileSystemErrors

**Comment by lramos15** [comment]
> This may be user facing in certain cases, so what is the reason for making the error name rather cryptic by concatenating `ErrorNoTelemetry` and the original `name`? I would rather use something that is more generic, since I believe the main intent is to implement `isErrorNoTelemetry` properly:

Yeah we can go with anything here. I think the main goal with sticking with `ErrorNoTelemetry` is to also try and clean it wherever it shows up

File: src/vs/base/common/errors.ts (hunk 2 above)
Comment (bpasero) [reply src/vs/base/common/errors.ts:252]: Yeah, so file system errors flow from processes into the ext host but also from extensions back into the renderer.

**Comment by lramos15** [comment]
I went with a very simple solution here as further exploration looked like it was going to break things down the line. I believe FileSystemErrors are the most complex case so my fix is ok but I wish everything could go through `ErrorNoTelemetry` to make it clear. Will reinvestigate what we can do to unblock the custom error properties in the future but for now this helps cleanup our data pipeline by removing file system errors

File: src/vs/platform/telemetry/common/errorTelemetry.ts [hunk 3]
```diff
@@ -87,7 +88,8 @@ export default abstract class BaseErrorTelemetry {
 		}
 
 		// If it's the no telemetry error it doesn't get logged
-		if (ErrorNoTelemetry.isErrorNoTelemetry(err)) {
+		// TOOD @lramos15 hacking in FileOperation error because it's too messy to adopt ErrorNoTelemetry. A better solution should be found
+		if (ErrorNoTelemetry.isErrorNoTelemetry(err) || err instanceof FileOperationError) {
```
Comment (bpasero) [inline src/vs/platform/telemetry/common/errorTelemetry.ts:92]: I think this is actually not so bad, and quite elegant.

**Review by bpasero — APPROVED** [review]

File: src/vs/platform/telemetry/common/errorTelemetry.ts (hunk 3 above)
Comment (lramos15) [reply src/vs/platform/telemetry/common/errorTelemetry.ts:92]: The downside here is it makes it unclear when looking at the error events whether or not they will be recorded. I liked the error producer deciding rather than us as the consumer
